### ISSUE TITLE
[Feature][Ascend NPU] Boogu-Image four-model BF16 support and Hybrid W8A8

### ISSUE THREAD

**Issue by QwertyJack** [vllm-project/vllm-omni#6787]
## Summary

This issue tracks the Ascend contribution for the Boogu-Image roadmap in #6665.
The scope has two milestones:

1. Complete, directly-served Ascend NPU BF16 support for Base, Turbo, Edit, and Edit-Turbo.
2. A separate Hybrid W8A8 implementation based on a reviewed checkpoint and loader contract.

Owner: @QTwzc
Related roadmap: #6665

## M1: Four-model BF16 support

Target stack for the first validation pass: CANN 9.0.0, PyTorch 2.10, and torch-npu 2.10, subject to the supported vLLM-Omni/vLLM-Ascend release pairing.

- [ ] Base T2I on Ascend with the native vLLM-Omni pipeline.
- [ ] Turbo T2I with the upstream four-step decoupled-DMD path, not only a class alias.
- [ ] Edit TI2I with a real reference-image request.
- [ ] Edit-Turbo TI2I with four steps and guidance scale 1.0.
- [ ] Real-valued RoPE, coordinated with #6571.
- [ ] Native NPU GQA in the shared SDPA backend, including full-QK masks and a CPU/fallback path.
- [ ] NPU RMSNorm and SwiGLU through the existing Omni diffusion layer dispatch, coordinated with #6649.
- [ ] Direct `vllm serve` smoke and real-weight first-request validation.
- [ ] NPU unit/numerics tests, A/B latency and memory numbers, perf baseline updates, supported-model and recipe updates, and expansion e2e rows.

### Pipeline identity contract

The published metadata currently declares:

- Base: `BooguImagePipeline`
- Turbo: `BooguImageTurboPipeline`
- Edit: `BooguImagePipeline`
- Edit-Turbo (main and `hotfix-1k-20260708`): `BooguImagePipeline`

Turbo semantics must therefore be selected by an explicit, auditable model/revision mapping or a pipeline override. Directory-name inference is not acceptable. The Edit-Turbo hotfix revision must reach every component loader.

## M2: Hybrid W8A8

The local validated candidate is `dit-hybrid-w8a8-v2` plus the official Ascend Qwen3-VL W8A8 QuaRot checkpoint. The current evidence is on 64 GB Ascend 910 only; physical 32 GB 910B4 validation is still required before claiming 32 GB support.

The implementation will be preceded by an RFC and will define:

- component routing and lazy/load-time materialization;
- DiT W8A16 attention/modulation;
- DiT FFN static shared-input W8A8 plus fused SwiGLU and dynamic per-token W8A8 down projection;
- packed QKV/gate-up checkpoint mapping after #6649;
- complete Qwen QuaRot restoration (rotated embeddings/norms, text linears, visual bridges, and final inverse transform);
- four-model quality, latency, memory, and real-checkpoint e2e validation.

This is not equivalent to the existing dynamic-activation `DiffusionInt8Config`, and it must not be silently exposed as generic `int8`.

## Dependencies and non-goals

- Depend on #6571, #6699/#6682, and #6701 where their changes are accepted; resolve overlapping Boogu files manually during integration.
- Coordinate #6649 so the shared `SiluAndMul` path is used where appropriate.
- FP8 remains #6677; Cache-DiT remains #6679; SP/CFG/TP/DP/HSDP remain their existing issues/PRs.
- #6610 (standalone torch-npu diffusion backend) is a separate platform decision and is not bundled into the first model PR.
- AMD/Intel support and offline upstream demo scripts are out of scope.

## Acceptance checklist

- [ ] Four real checkpoints produce finite, non-empty images through the vLLM-Omni API; both T2I and TI2I are covered.
- [ ] NPU numerics are compared against a BF16 reference with fixed request parameters and documented tolerances.
- [ ] A/B performance and peak allocated/reserved memory are recorded; `tests/dfx/perf/tests/test_boogu_image_*.json` is updated.
- [ ] `docs/models/supported_models.md`, `recipes/Boogu/Boogu-Image.md`, and expansion e2e coverage are updated per landed capability.
- [ ] Dummy and real-weight results are reported separately; startup-only evidence is not accepted.
- [ ] Physical 32 GB 910B4 is validated before the W8A8 milestone is marked complete.

**Comment by QwertyJack** [comment]
First shared-backend PR opened: #6788 (draft). It preserves compressed K/V for native NPU SDPA GQA and adds CPU numerical/dispatch/full-QK-mask coverage. Real-NPU numerical and Boogu end-to-end A/B validation remain required before it is marked ready.

**Comment by QwertyJack** [comment]
Quantization RFC opened as #6789. It records the validated Hybrid W8A8 v2 boundary, rejects direct post-load module replacement, and asks maintainers to decide the quant method/manifest and vllm-ascend versus standalone torch-npu ownership before implementation.

**Comment by QwertyJack** [comment]
Current integration status:

- `#6571 + #6649 + #6699 + #6701 + #6788` are combined on `codex/boogu-m1-integration` with the local model-identity/shared-layer fixes.
- Focused combined tests: `70 passed` for Boogu registry/pipeline + SDPA/GQA, and `16 passed` for the Boogu transformer.
- The standalone #6788 branch passes its focused SDPA/GQA suite (`7 passed`); all public checks are green.
- Entry/alignment collection is not counted as passing in this lightweight environment because its installed vLLM source lacks `CoreEngineLaunch`, which current Omni imports.
- All 16 visible Ascend 910 devices are still occupied by unrelated `VLLMWorker_DP` processes, so real-NPU GQA and four-model service acceptance remain pending. Those processes were not stopped.

I also reported the `freqs_cis`/`freqs_real` combination regression on #6699 and the official-model/local-path identity contract on #6698.

**Comment by QwertyJack** [comment]
M2 construction progress: #6791 is now open as a draft stacked on #6649. It wires component-scoped quant configs and stable Boogu DiT prefixes without claiming Hybrid W8A8 runtime support.

All changed-file pre-commit hooks pass locally. Real-NPU and four-model acceptance remain pending because all visible devices are still occupied by unrelated `VLLMWorker_DP` processes.

**Comment by QwertyJack** [comment]
M1 real-NPU bring-up update (2026-08-30): all four published BF16 checkpoints now complete a real-weight API request on one Ascend 910_9382 device with the integration stack `#6571 + #6649 + #6699 + #6701 + #6788 + #6792` plus the focused model-identity fixes.

| Model | Request path | Settings | Result | API E2E |
|---|---|---:|---:|---:|
| Base | `/v1/images/generations` | 50 steps, guidance 4.0, 1024x1024 | HTTP 200 | 51.75 s |
| Turbo | `/v1/images/generations`, native DMD | 4 steps, guidance 1.0, 1024x1024 | HTTP 200 | 2.29 s |
| Edit | `/v1/images/edits`, real reference image | 50 steps, guidance 4.0, 1024x1024 | HTTP 200 | 117.84 s |
| Edit-Turbo | `/v1/images/edits`, native DMD | 4 steps, guidance 1.0, 1024x1024 | HTTP 200 | 5.22 s |

All four responses contained one decodable, non-empty RGB PNG. SHA256:

```text
Base        ee45788cd7e38d53aa8bd4cfba60528cc37b3cf6424608c66a50fe870f8e153c
Turbo       69499a9b7ce1dd59a17c0459fc16b2477b31f89ea2d4d79587cd1f9143efe190
Edit        ed36ad1fc7f621ede81d5fb97cdef1da39e1a736f9b599d513d334ed77e2840d
Edit-Turbo  3772ca901fc50fb3841c7d7a1861cbaf40ad591015933f4ad57813da87f03d8b
```

The Edit request replaced the background while preserving the foreground subject in this smoke case. This is functional/visual-smoke evidence only, not numerical or quality sign-off; fixed BF16 reference trajectory/pixel/semantic comparison is still required.

Edit-Turbo provenance was checked rather than inferred from its directory name. Its eight local MLLM/DiT/VAE LFS SHA256 OIDs match the current HF `main`; the complete runtime file tree for current `main` and tag `hotfix-1k-20260708` is identical. Because the published model index still declares `BooguImagePipeline`, the local checkpoint was served with an explicit `--model-class-name BooguImageTurboPipeline`. The four-step request therefore exercised #6699's `_is_turbo=True` predict-x0/renoise path rather than regular Edit with a smaller step count.

Shared native-GQA evidence for #6788 was also collected on the same NPU. At Boogu's `28 Q heads / 7 KV heads / head_dim 120` shape (`seq=512`, full-QK bool mask), compressed-KV native GQA and explicit K/V expansion were bit-identical in this probe (`max/mean abs error = 0`). Mean latency over 20 iterations was 0.279 ms native versus 0.308 ms expanded, while avoiding 4.92 MiB of Python-side K/V materialization per call.

Important environment limits:

- vLLM `2cf0a691`, vLLM-Ascend `e6a133f710`, vLLM-Omni integration `3307f5c932`; PyTorch 2.10.0 + torch-npu 2.10.0.post2; CANN 9.0.1.
- vLLM-Ascend at this commit declares torch-npu post4, triton-ascend 3.2.2, and FastAPI `<0.124`; this test venv has post2, triton-ascend 3.2.1, and FastAPI 0.136.3.
- vLLM-Ascend was installed with `COMPILE_CUSTOM_KERNELS=0` after the custom-kernel CMake build rejected the compiler's default C++ level. This proves the Boogu diffusion path on the Python/operator stack; it is not a release-image sign-off.
- Requests succeeded, but Ctrl-C shutdown and ordinary torch-npu pytest finalization can end with `corrupted size vs. prev_size`. The real-GQA assertion printed `1 passed` before the process aborted with exit 134. A small standalone probe used `os._exit(0)` after synchronization to isolate the operator result from that finalizer problem.

I am therefore keeping M1 open. Next gates are: land/coordinate the dependency PRs, repeat this matrix in a supported vLLM/vLLM-Ascend release image, add fixed-reference numerical/quality evidence, and add the NPU recipe/e2e/perf baseline without duplicating the owners of #6571/#6649/#6699/#6701.

**Comment by QTwzc** [comment]
Hi @QwertyJack, @hsliuustc0106, and @Gaohan123,

I can take the M1 validation. Please assign #6787 to @QTwzc.

I have prepared the quay.io/ascend/vllm-ascend:v0.22.1rc1-a3 environment with CANN 9.0.0, PyTorch 2.10, and torch-npu 2.10 on our Ascend 910C machine. Before I begin, I’d like to confirm two points:

Could you confirm the exact integration head/commits to reproduce? Should I use codex/boogu-m1-integration, or reconstruct it from the dependent PRs (including #6792) plus the local model-identity fixes?
Do we also need a validation run on CANN 9.0.1 for final sign-off? The current smoke results used 9.0.1, while this issue specifies 9.0.0 for the first validation pass.

I’ll report the four-model real-weight smoke, numerics, repeated latency/memory results, and shutdown status separately. I’ll wait for the #6789 design decision before starting M2.

**Comment by QTwzc** [comment]
Hi @QwertyJack, @hsliuustc0106, and @Gaohan123,

Before proceeding with the M1 work, I would like to confirm a proposed adjustment to the validation baseline.

The M1 first-pass target specifies CANN 9.0.0, PyTorch 2.10, and torch-npu 2.10, subject to a supported vLLM-Omni/vLLM-Ascend pairing. According to the public compatibility matrix, this stack maps to vLLM 0.22.1 and vLLM-Ascend 0.22.1rc1.

However, the vLLM-Omni 0.22 release predates the native Boogu pipeline, while the current vLLM-Omni main containing the Boogu implementation—and targeted by the recent M1-related PRs—depends on newer vLLM APIs. Backporting the current Boogu/M1 code to the 0.22 release line would therefore require substantial compatibility work before model validation can begin.

To avoid validating against a pre-Boogu Omni baseline, I plan to perform the initial four-model BF16 integration and related-PR validation on Ascend 910C using CANN 9.1.0 and the aligned v0.28 stack. The exact image digest and repository SHAs will be recorded, and all results will be explicitly labeled as CANN 9.1 validation rather than evidence for the original CANN 9.0 first-pass target.

Please let me know if this adjusted validation baseline is acceptable. If M1 must still be validated on CANN 9.0, please advise whether there is an intended integration branch or a specific vLLM-Omni/vLLM/vLLM-Ascend version combination for that stack.

**Comment by QTwzc** [comment]
Hi @QwertyJack @hsliuustc0106,

Regarding the M1 item **“NPU RMSNorm and SwiGLU through the existing Omni diffusion layer dispatch, coordinated with #6649,”** we are working on SwiGLU fusion for Boogu-Image-Turbo and would like to confirm the implementation direction.

In our current checkout, Boogu’s FFN uses:

```python
F.silu(gate.float()).to(gate.dtype) * up
```

The shared `SiluAndMul` already provides an NPU implementation through `torch_npu.npu_swiglu`. However, #6649 explicitly retains the original FP32 SwiGLU for numerical parity after merging the gate/up projections.

For an initial, isolated experiment, we plan to keep the separate gate/up projections and call the shared `SiluAndMul` with `torch.cat([gate, up], dim=-1)`. This would let us evaluate numerical differences and the net latency benefit, including concatenation overhead. The intended integration would then reuse #6649’s merged gate/up output to eliminate the concatenation.

Could you advise on two points?

1. Would a separate SwiGLU integration PR under this issue, coordinated with #6649 and using its merged projection layout, be appropriate?
2. Is the original FP32 intermediate computation and cast-before-multiply behavior a strict numerical requirement? Or could the fused implementation be accepted with agreed numerical tolerances and real-weight numerical and image-quality validation?

We will first provide numerical comparisons, trace evidence of the fused execution path, and profiler-off latency A/B results for Turbo, followed by regression validation for the other three models.
